bee-protocol/src/worker/propagator.rs

```diff
@@ -48,50 +49,66 @@ async fn propagate<B: StorageBackend>(
+            // we can simply unwrap. We also try to minimise unnecessary Tangle API calls if - for example - the
+            // parent in question turns out to be a SEP.
+
+            // Determine OTRSI/YTRSI of parent1
+            let (parent1_otrsi, parent1_ytrsi) = match tangle.get_solid_entry_point_index(&parent1) {
+                Some(parent1_sepi) => (IndexId(parent1_sepi, parent1), IndexId(parent1_sepi, parent1)),
+                None => match tangle.get_metadata(&parent1).await {
+                    Some(parent1_md) => (parent1_md.otrsi().unwrap(), parent1_md.ytrsi().unwrap()),
+                    None => continue,
+                },
+            };
+
+            // Determine OTRSI/YTRSI of parent2
+            let (parent2_otrsi, parent2_ytrsi) = match tangle.get_solid_entry_point_index(&parent2) {
+                Some(parent2_sepi) => (IndexId(parent2_sepi, parent2), IndexId(parent2_sepi, parent2)),
+                None => match tangle.get_metadata(&parent2).await {
+                    Some(parent2_md) => (parent2_md.otrsi().unwrap(), parent2_md.ytrsi().unwrap()),
+                    None => continue,
+                },
+            };
+
+            // Derive child OTRSI/YTRSI from its parents.
+            // Note: The child inherits oldest (lowest) otrsi and the newest (highest) ytrsi from its parents.
+            let child_otrsi = min(parent1_otrsi, parent2_otrsi);
+            let child_ytrsi = max(parent1_ytrsi, parent2_ytrsi);
+
+            let ms_index_maybe = tangle
                 .update_metadata(&message_id, |metadata| {
-                    // OTRSI/YTRSI values need to be set before the solid flag, to ensure that the
-                    // MilestoneConeUpdater is aware of all values.
-                    metadata.set_otrsi(new_otrsi);
-                    metadata.set_ytrsi(new_ytrsi);
-                    metadata.solidify();
+                    // The child inherits the solid property from its parents.
+                    metadata.set_solid();
 
                     if metadata.flags().is_milestone() {
                         Some(metadata.milestone_index())
                     } else {
+                        metadata.set_otrsi(child_otrsi);
+                        metadata.set_ytrsi(child_ytrsi);
                         None
                     }
                 })
                 .await
                 .expect("Failed to fetch metadata.");
 
+            // Try to propagate as far as possible into the future.
             if let Some(msg_children) = tangle.get_children(&message_id).await {
                 for child in msg_children {
                     children.push(child);
                 }
             }
 
-            let _ = tip_tx.send((*message_id, parent1, parent2, index)).await;
+            // Send child to the tip pool.
+            let _ = tip_tx.send((*message_id, parent1, parent2, ms_index_maybe)).await;
```

Comment:
maybe a `if let Err(err) = tip_tx.send((*message_id, parent1, parent2, ms_index_maybe)).await {}` as we do we other channels would be better here

Reply:
sure, I left that line untouched. Will change it.

Reply:
actually not entirely untouched :joy:

Reply:
added a warning